partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt

```diff
@@ -26,22 +36,114 @@ internal class ExprCallDynamic(
     private val args: Array<Operator.Expr>
 ) : Operator.Expr {
 
-    private val candidateIndex = CandidateIndex.All(candidates)
+    private val paramIndices: IntRange = args.indices
+    private val paramTypes: List<List<PType>> = this.candidates.map { candidate -> candidate.fn.signature.parameters.map { it.type } }
+    private val paramFamilies: List<List<CoercionFamily>> = this.candidates.map { candidate -> candidate.fn.signature.parameters.map { family(it.type.kind) } }
+    private val cachedMatches: MutableMap<List<PType>, Int> = mutableMapOf()
```

Comment:
Interesting that this is a `MutableMap<List<PType>>` rather than something like `MutableMap<List<PType.Kind>>`.

Reply:
It has to do with our function resolution. We follow PostgreSQL for exact matches -- which take into account parameters. So, in one dispatch (say argument of type `DECIMAL(2,0)`) could be matched to one candidate, while another dispatch (say argument `DECIMAL(3,0)`) could be matched to another candidate.

If we change our modelling of function signatures, then this can change as well.